[SYCL] The sycl_kernel_entry_point attribute.

tahonermann opened the pull request:
The `sycl_kernel_entry_point` attribute is used to declare a function that defines a pattern for an offload kernel to be emitted. The attribute requires a single type argument that specifies the type used as a SYCL kernel name as described in [section 5.2, "Naming of kernels", of the SYCL 2020 specification](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:naming.kernels).

Properties of the offload kernel are collected when a function declared with the `sycl_kernel_entry_point` attribute is parsed or instantiated. These properties, such as the kernel name type, are stored in the AST context where they are (or will be) used for diagnostic purposes and to facilitate reflection to a SYCL run-time library. These properties are not serialized with the AST but are recreated upon deserialization.

The `sycl_kernel_entry_point` attribute is intended to replace the existing `sycl_kernel` attribute which is intended to be deprecated in a future change and removed following an appropriate deprecation period. The new attribute differs in that it is enabled for both SYCL host and device compilation, may be used with non-template functions, explicitly indicates the type used as the kernel name type, and will impact AST generation.

This change adds the basic infrastructure for the new attribute. Future changes will add diagnostics and new AST support that will be used to drive generation of the corresponding offload kernel.

tahonermann committed:
[SYCL] The sycl_kernel_entry_point attribute.

The `sycl_kernel_entry_point` attribute is used to declare a function that
defines a pattern for an offload kernel to be emitted. The attribute requires
a single type argument that specifies the type used as a SYCL kernel name as
described in section 5.2, "Naming of kernels", of the SYCL 2020 specification.

Properties of the offload kernel are collected when a function declared with
the `sycl_kernel_entry_point` attribute is parsed or instantiated. These
properties, such as the kernel name type, are stored in the AST context where
they are (or will be) used for diagnostic purposes and to facilitate reflection
to a SYCL run-time library. These properties are not serialized with the AST
but are recreated upon deserialization.

The `sycl_kernel_entry_point` attribute is intended to replace the existing
`sycl_kernel` attribute which is intended to be deprecated in a future change
and removed following an appropriate deprecation period. The new attribute
differs in that it is enabled for both SYCL host and device compilation, may
be used with non-template functions, explicitly indicates the type used as
the kernel name type, and will impact AST generation.

This change adds the basic infrastructure for the new attribute. Future
changes will add diagnostics and new AST support that will be used to drive
generation of the corresponding offload kernel.

tahonermann commented:
This is the first in a series of PRs to provide core language support for SYCL-aware host and device compilation and support for SYCL library implementations. The following links present a preliminary set of changes for future PRs that depend on this one. Some of these need a little work yet (e.g., some of the diagnostics need a bit of tweaking), but will be submitted in order as dependencies are accepted.

- PR 1: [SYCL] The sycl_kernel_entry_point attribute. (this PR)
- PR 2: [\[SYCL\] Diagnostics for the sycl_kernel_entry_point attribute.](https://github.com/tahonermann/llvm-project/compare/sycl-upstream-fe-pr1...sycl-upstream-fe-pr2)
- PR 3: [\[SYCL\] AST support for SYCL kernel entry point functions.](https://github.com/tahonermann/llvm-project/compare/sycl-upstream-fe-pr2...sycl-upstream-fe-pr3)
- PR 4: [\[SYCL\] Offload kernel code generation for SYCL kernel entry point functions.](https://github.com/tahonermann/llvm-project/compare/sycl-upstream-fe-pr3...sycl-upstream-fe-pr4)
- ...
- PR X: [\[SYCL\] Remove the sycl_kernel attribute.](https://github.com/tahonermann/llvm-project/compare/sycl-upstream-fe-pr4...sycl-upstream-fe-rm-sycl_kernel-attr)

The last one listed above is to remove the existing `sycl_kernel` attribute pending further discussion. It is believed that there are no users of that attribute (other than Intel) and that it might be possible to remove it without a deprecation period. We'll follow up with an RFC for that at a later date.

erichkeane commented on clang/include/clang/AST/SYCLKernelInfo.h:
```diff
@@ -0,0 +1,47 @@
+      KernelEntryPointDecl(KernelEntryPointDecl)
+  {}
+
+  CanQualType GetKernelNameType() const {
```
```suggestion
  CanQualType getKernelNameType() const {
```
etc

tahonermann replied:
Thank you, fixed.

erichkeane commented on clang/include/clang/AST/SYCLKernelInfo.h:
```diff
@@ -0,0 +1,47 @@
+    return KernelNameType;
+  }
+
+  const FunctionDecl* GetKernelEntryPointDecl() const {
```
```suggestion
  const FunctionDecl* getKernelEntryPointDecl() const {
```

tahonermann replied:
Thank you, fixed.

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,64 @@ The SYCL kernel in the previous code sample meets these expectations.
+def SYCLKernelEntryPointDocs : Documentation {
+  let Category = DocCatFunction;
+  let Content = [{
+The ``sycl_kernel_entry_point`` attribute specifies that a function definition
```
This first sentence is... very difficult to understand/self referential. Reword please.

tahonermann replied:
Yeah. I rewrote a bunch of the doc. Hopefully better now.

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,64 @@ The SYCL kernel in the previous code sample meets these expectations.
+kernel on a device. The function's parameters define the parameters to the
+offload kernel.
+
+The attribute requires a single type argument that specifies a class type that
```
I like this development.  The work that is done in `sycl_kernel` to pull this out of the template argument was always problematic to me.

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,64 @@ The SYCL kernel in the previous code sample meets these expectations.
+functions that meet the following requirements:
+
+- Has a ``void`` return type.
+- Is not a variadic function.
```
Should probably clarify you mean a C variadic, as there isn't really a problem with variadic function template.

tahonermann replied:
Thank you, fixed.

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,64 @@ The SYCL kernel in the previous code sample meets these expectations.
+- Has a ``void`` return type.
+- Is not a variadic function.
+- Is not a coroutine.
+- Is not defined as deleted or as defaulted.
```
Why not defaulted?  I could definitely see a reasonable use case for this, though less so due to the member-function limitation.

tahonermann replied:
I haven't been able to imagine such a use case. Can you explain further how you envision this being useful? I'm inclined toward prohibiting use with defaulted functions for now; we can always relax support later if a use case emerges.

tahonermann replied:
I'm resolving this thread regarding prohibition of defaulted functions. Erich, if you have further thoughts on it, please unresolve.

erichkeane replied:
I missed your response/question here, and honestly cannot remember what I was thinking about here.

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,64 @@ The SYCL kernel in the previous code sample meets these expectations.
+is required for each function declared with the attribute. The attribute may
+not first appear on a declaration that follows a definition of the function.
+
+The attribute appertains only to non-member functions and static member
```
I don't have a good suggestion here, but `non-member functions` and `static member functions` reads weird, as the 1st says "no member functions" the 2nd says "some member functions".  Perhaps invert this to say something like:

appertains to any functions except non-static member functions that meet the following requirements...

Also, what about lambdas?

tahonermann replied:
I reworded by moving the restriction on non-static member functions into the bullet list. I also added constructors and destructors.

Historically, lambdas would be prohibited by the restrictions against non-static member functions. We now have lambda expressions that produce a closure type with a static fiunction call operator. I think those should be ok.

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,64 @@ The SYCL kernel in the previous code sample meets these expectations.
+- Is not a variadic function.
+- Is not a coroutine.
+- Is not defined as deleted or as defaulted.
+- Is not declared with the ``constexpr`` or ``consteval`` specifiers.
```
Why no constexpr?  I get no consteval, but constexpr seems reasonable, it should not do offload during constant evaluation (obviously), but that shouldn't prevent being marked as constexpr.  This is particularly problematic in cases of lambdas that end up being implicitly constexpr, which we might want to mark, but it won't be clear this is the case.

tahonermann replied:
Hmm, maybe? I guess I'm not opposed, but it seems strange. There is no possibility of the offload kernel entry point being invoked in a constant evaluation context. The possibility of `constexpr` being implicitly added is a good point, but I don't really see this attribute being used with (static) lambdas either. I'm inclined to wait and see if use cases arise.

tahonermann replied:
Erich, we have another conversation going regarding `constexpr` support. I'm going to resolve this one and we can continue any further discussion on the other one.

erichkeane commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,64 @@ The SYCL kernel in the previous code sample meets these expectations.
+- Is not declared with the ``constexpr`` or ``consteval`` specifiers.
+- Is not declared with the ``[[noreturn]]`` attribute.
+
+This attribute is intended for use in the implementation of SYCL run-time
```
Is this something we should enforce?  That is, make this legal only in a `system-header` file?

tahonermann replied:
I don't think so. That would tend to bias against SYCL library implementations that aren't provided with LLVM/Clang or distributed with an OS. I don't know that such implementations are likely to appear, but I'm not inclined towards adding roadblocks for them.

tahonermann replied:
Erich, I'm going to consider this conversation about constraining use of the attribute to system headers resolved. If you have further thoughts on it, please unresolve.

erichkeane replied:
Again, no idea how I missed the response.  I think your answer here is sensible.

erichkeane commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -14296,6 +14296,31 @@ void ASTContext::getFunctionFeatureMap(llvm::StringMap<bool> &FeatureMap,
+
+void ASTContext::registerSYCLEntryPointFunction(FunctionDecl *FD) {
+  assert(!FD->isInvalidDecl());
+  assert(!FD->isDependentContext());
```
Do you mean to check 'template-kind' here?

tahonermann replied:
No. At least, I don't think so. The intent is to reject any declaration that is somehow templated. A static member function of a class template (as opposed to its instantiated declaration in an instantiated specialization) shouldn't be registered.

tahonermann replied:
Erich, I'm assuming my response satisfied your question about checking for a dependent context as opposed to a template kind here. I'm guessing your thoughts about template kind were motivated by an expectation that the attribute would only appertain to function templates; a restriction we dropped as discussed elsewhere.

erichkeane replied:
Again, no idea how I missed this... Github seems to hide comments from me at a certain point.  

I think that was probably true.  As a nit, we check this with `isTemplated` IIRC, but I don't have a good idea of when this and `isDependentContext` would differ, so I don't think there is anything to do here?

tahonermann replied:
I'm struggling with GitHub hiding comments too. The only reliable way I've found to make sure I'm following up on all of them is to stay in the conversation tab, scroll slowly from the top, expanding collapsed sections and resolved conversations as I go. Comments on "outdated" files seem to be implicitly hidden or at least aren't shown on the files tab at all. There have been numerous times where I know a conversation happened and it takes me a long time to find it again. Page searches don't identify text in collapsed sections (at least not always). Quite frustrating. I still miss Phabricator. Anyway, I digress.

Thank you for pointing me to `isTemplated()`. It uses `isDependentContext()`, but has special handling for friend and local function declarations. I think this should be using `isTemplated()`, so I'll change that and look to see if test updates are appropriate. Good catch.

erichkeane replied:
Yeah, I'm finding that sort of thing as well, which is frustrating, particularly because 'code view' is what i always WANT anyway (and doesn't result in you getting 300 emails as I respond to messages, as I can do them as reviews).  Its like no one at github actually uses this tool...

Cool, glad you looked into it!  I knew I looked into it and preferred it at one point, but didn't recall why (and looking it up on my phone seemed.... onerous).

tahonermann replied:
I changed the code to use `isTemplated()` instead of `isDependentContext()`, so resolving this conversation now.

I wasn't able to identify a test case where it made a difference, but the change seems right regardless. I added a test for a dependent friend function for the next PR. The test looks like:
```
template<typename KNT>
struct S7 {
  [[clang::sycl_kernel_entry_point(KNT)]]
  friend void ok7(S7) {}
};
void test_ok7() {
  ok7(S7<KN<7>>{});
}
```

erichkeane commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -14296,6 +14296,31 @@ void ASTContext::getFunctionFeatureMap(llvm::StringMap<bool> &FeatureMap,
+  auto IT = SYCLKernels.find(KernelNameType);
+  if (IT != SYCLKernels.end()) {
+    if (!declaresSameEntity(FD, IT->second.GetKernelEntryPointDecl()))
+      llvm::report_fatal_error("SYCL kernel name conflict");
```
This absolutely should not be a fatal error here, this should be a diagnosed Sema-esque error.

tahonermann replied:
Very much agreed. Diagnostics will be available in the next PR. There are still a few issues we're working to fix before that PR is submitted though. Our WIP is staged [here](https://github.com/tahonermann/llvm-project/compare/sycl-upstream-fe-pr1...tahonermann:llvm-project:sycl-upstream-fe-pr2). Feel free to take a look and offer any comments here; I'll then address them before that PR is submitted.

tahonermann replied:
This isn't intended to be a temporary fatal error. Diagnostics should be (and are in the [staged changes for the next PR](https://github.com/tahonermann/llvm-project/compare/sycl-upstream-fe-pr1...tahonermann:llvm-project:sycl-upstream-fe-pr2)) issued before registration is attempted. This is intended to catch future bugs where diagnostics are missing. I chose `report_fatal_error()` over `assert()` to ensure such problems are revealed in release builds since a conflicting registration would indicate a bad situation that should never be permitted to escape until run-time.

erichkeane replied:
That is just not something we do in Clang, we use fatal error for situations where we SHOULD have handled it, but don't (yet).  I believe this to be the wrong choice.

AaronBallman replied:
I would not expect a function prefixed with `register` to issue a diagnostic or report a fatal error, I would expect that to be handled by the caller, which is what it looks like the follow-up patch does.

I would expect this to be an assert (and we probably should write some documentation in the internals manual about this). Generally speaking, we prefer:

* issuing a user-facing diagnostic when the user has done something wrong
* using an assert when we think a situation exhibits a Clang developer holding an API incorrectly but we have a reasonable recovery path
* using a fatal error when we think a situation exhibits a Clang developer holding an API incorrectly and there's no possible way to recover from it
* using llvm_unreachable for code paths which should never be reached

This situation seems like an assert because 1) someone called the register function when they shouldn't have, the caller should have reported this as an error to the user (and that's what's happening in a follow-up), and 2) we have a reasonable recovery path, which is to not register the entrypoint.

erichkeane replied:
Aaron: yeah, that reflects my thoughts on this within some nits on that list that are unimportant to the discussion.

tahonermann replied:
This PR doesn't provide enough functionality to demonstrate what would happen if Aaron's suggested recovery path was implemented. Nor do we have enough staged changes to easily perform a full end-to-end test. However, I was able to experiment with the set of changes we currently have staged [here](https://github.com/tahonermann/llvm-project/compare/main...tahonermann:llvm-project:sycl-upstream-fe-pr4). To do so, I removed the disputed call to `llvm::report_fatal_error()` (along with another one currently present in the linked staged changes) and the diagnostics that protect against that call and compiled the following test case.
```
template<typename KN, typename KT>
[[clang::sycl_kernel_entry_point(KN)]]
void skep(KT k) { k(); }
void f() {
  skep<struct K>([]{});
  skep<struct K>([]{});
}
```

Compilation was successful (it shouldn't have been of course; it succeeded because the diagnostics were elided). I then checked what symbols were emitted for device compilation. A little tangent first. As noted in the documentation added by this PR, ODR-use of a `sycl_kernel_entry_point` attributed function triggers emission of an offload kernel entry point function. The staged changes emit that function named as though it were a specialization of a function template named `__sycl_kernel_caller` with a single type template argument corresponding to the SYCL kernel name. Going into this experiment, I expected to end up with some kind of error or failed assertion related to an attempt to generate a duplicate symbol. But instead, some kind of function versioning kicked in.
```
$ nm t2.o
...
0000000000000020 T _Z20__sycl_kernel_callerIZ1fvE1KEvv
0000000000000040 T _Z20__sycl_kernel_callerIZ1fvE1KEvv.1
...
```

I'm not in a position to test with an actual SYCL run-time library, so I can only speculate here, but this looks like a scenario that I was worried about. Two entry point functions were emitted, but the SYCL run-time would only resolve the kernel name to one of them with the result that the wrong entry point would be silently invoked for some cases (had actual SYCL kernel invocation functions been used in the example). That would not be fun to debug.

We certainly can (and will) change our staged code to avoid the function versioning that is happening here; we should detect the situation and emit some kind of duplicate definition diagnostic or similar. With that kind of backup in place, I would feel more comfortable replacing the call to `llvm::report_fatal_error()` with an assert. But this is an example of why I have felt that we should do more to prevent such Bad Things from happening.

erichkeane replied:
I would expect us to use the normal diagnostics engine on the 2nd to last line there to diagnose that.  I think we did something similar in `intel/llvm` that should be what you're looking for.  However, any such diagnostic is going to be imperfect cross-TU, but `report_fatal_error` cant do that anyway.  Perhaps there is a need for a `checkSYCLEntryPointFunction` that does this look up vs previous ones.

tahonermann replied:
Agreed. That diagnostic is implemented in the staged changes for the next PR; it is one of the ones I had to disable to produce the (surprising to me) symbol versioning result I described. Again, my goal here was to reduce the negative consequences of a missed diagnostic; I think an ICE is preferred over the possibility of an incorrect kernel being called at run-time.

I'm following up internally to do something about the implicit symbol versioning issue. We should have that possibility eliminated before the code generation PR is submitted.

> However, any such diagnostic is going to be imperfect cross-TU, but report_fatal_error cant do that anyway.

Indeed. We'll need to rely on linkers rejecting duplicate symbol definitions for that.

tahonermann replied:
I pushed changes to replace the call to `llvm::report_fatal_error()` with `assert()` following discussion with other colleagues. I still disagree with this direction. As a user, I would much prefer that the compiler fail catastrophically than continue merrily along generating a binary that doesn't behave as expected (In these days of safety consciousness, we would really rather continue into UB land than abort? Really?) . I acknowledge such eventualities will only occur if we fail to properly diagnose elsewhere, so the stakes are hopefully not so high. Still, I reserve the right to be unhappy and say, "I told you so" while vigorously wagging a finger if we end up getting a bug report detailing someone's difficult experience debugging why their application was mysteriously invoking the wrong SYCL kernel. I'm now resolving this conversation.

erichkeane commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -14296,6 +14296,31 @@ void ASTContext::getFunctionFeatureMap(llvm::StringMap<bool> &FeatureMap,
+    if (!declaresSameEntity(FD, IT->second.GetKernelEntryPointDecl()))
+      llvm::report_fatal_error("SYCL kernel name conflict");
+  } else {
+    SYCLKernels.insert_or_assign(
```
Why `insert_or_assign`?  We already know it doesn't exist.  Or is this `else` on the wrong `if` here?

tahonermann replied:
Thank you. I think the "or_assign" part was leftover from before some refactoring. Fixed.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -198,3 +198,12 @@ void SemaSYCL::handleKernelAttr(Decl *D, const ParsedAttr &AL) {
+  ParsedType PT = AL.getTypeArg();
+  TypeSourceInfo *TSI = nullptr;
+  (void)SemaRef.GetTypeFromParser(PT, &TSI);
+  assert(TSI && "no type source info for attribute argument");
```
Have we already checked that the user gave us a valid name somewhere?

tahonermann replied:
Erich, please unresolve this conversation about valid type names if you have further concerns.

erichkeane commented on clang/test/SemaSYCL/sycl-kernel-entry-point-attr-grammar.cpp:
```diff
@@ -0,0 +1,137 @@
+// RUN: %clang_cc1 -triple x86_64-linux-gnu -std=c++17 -fsyntax-only -fsycl-is-device -verify %s
```
I don't think I see tests for all of the things in your documented list.

tahonermann replied:
Those tests are included with the diagnostic work that will be in the next PR. See [here](https://github.com/tahonermann/llvm-project/compare/sycl-upstream-fe-pr1...tahonermann:llvm-project:sycl-upstream-fe-pr2) for what has been completed. We still have some SFINAE issues to address (an invalid use of the attribute is triggering SFINAE and therefore affecting overload resolution) and I'm still working on a diagnostic for forward declarability of kernel names (for SYCL portability warnings; Clang won't actually require forward declarability since it won't depend on integration headers).

tahonermann replied:
Erich, I'm going to resolve this conversation regarding tests and diagnostics (planned for the next PR). Please unresolve if there is more you want to see in this PR.

erichkeane replied:
I think that is fine for me.  Again, sorry for not seeing this one :/  I would LOVE to see the list of 'things not diagnosed' documented/updated somewhere (so that we can check to make sure we get them all, and more importantly (hah!) I can do so while reviewing).  But I'm OK holding off on this.

tahonermann committed:
Fix clang-format complaints.

tahonermann committed:
Address code review feedback from Erich Keane.

tahonermann commented:
@erichkeane, I think I have addressed or replied to all of your comments. Another round of review feedback would be much appreciated!

@bader, as SYCL code maintainer, I would appreciate if you can take a look.

erichkeane commented:
> @erichkeane, I think I have addressed or replied to all of your comments. Another round of review feedback would be much appreciated!
> 
> @bader, as SYCL code maintainer, I would appreciate if you can take a look.

I didnt get a chance to get back to this during the week, but intend to after llvm-dev. By the time you got to responding, I'd context-switched it, so I'll have to spend some time getting back to it.

keryell reviewed (approved):
LGTM.
Thanks for the good documentation!

bader commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -14296,6 +14296,30 @@ void ASTContext::getFunctionFeatureMap(llvm::StringMap<bool> &FeatureMap,
   }
 }
 
+static SYCLKernelInfo BuildSYCLKernelInfo(ASTContext &Context,
```
`Context` parameter is unused. I suppose it might trigger a compiler warning. Can we drop this parameter?

tahonermann replied:
Oops, yes. It will be needed in future PRs, but should be added then. I'll remove it in this one.

tahonermann replied:
Done.

bader reviewed (approved):
LGTM, thanks!
I second @keryell: thank you for the good documentation.

tahonermann committed:
Address code review feedback from Alexey Bader.

tahonermann committed:
Fix clang-format complaints.

erichkeane commented on clang/lib/Serialization/ASTReaderDecl.cpp:
```diff
@@ -1155,6 +1155,15 @@ void ASTDeclReader::VisitFunctionDecl(FunctionDecl *FD) {
+  // If the declaration is a SYCL kernel entry point function as indicated by
+  // the presence of a sycl_kernel_entry_point attribute, register it so that
+  // associated metadata is recreated.
+  if (!FD->isInvalidDecl() && !FD->isDependentContext() &&
```
yeah, i'm quite sure now that `register..` should just accept/exit early in these cases.

tahonermann replied:
I went ahead and made the change to silently skip the registration for invalid or dependent function declarations and removed the corresponding checks at call sights. It isn't my preference, but I decided I don't feel strongly about it. Resolving this conversation.

erichkeane commented on clang/test/ASTSYCL/ast-dump-sycl-kernel-entry-point.cpp:
```diff
@@ -0,0 +1,144 @@
+void skep1() {
+}
+// CHECK: |-FunctionDecl {{.*}} skep1 'void ()'
+// CHECK: | `-SYCLKernelEntryPointAttr {{.*}} KN<1>
```
A lot of `CHECK` lines here that should probably be `CHECK-NEXT`.

tahonermann replied:
That is probably right. I'll take a closer look and update as appropriate.

tahonermann replied:
I changed `CHECK` lines to use `CHECK-NEXT` where applicable. There weren't all that many since the test avoids checking aspects of the AST that aren't particularly relevant. The goal is to check that the `SYCLKernelEntryPointAttr` nodes are correct; the rest are there to validate the correct context.

erichkeane replied:
Ah, thats fine then.  I was thinking at least with the highlighted line that ensuring it is associated with `skep1` would be particularly important, but of course skipping extra check lines is fine.

erichkeane reviewed (commented):
A handful of minor things, but overall I think this is alright.

keryell commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -407,7 +407,8 @@ def MicrosoftExt : LangOpt<"MicrosoftExt">;
 def CUDA : LangOpt<"CUDA">;
 def HIP : LangOpt<"HIP">;
-def SYCL : LangOpt<"SYCLIsDevice">;
+def SYCLHost : LangOpt<"SYCLIsHost">;
```
On the other hand SYCL is a trademark from Khronos, so if it can stay SYCL it is better from this perspective.
Otherwise people get confused and create more effort like https://github.com/llvm/llvm-project/pull/113060 :smiley:

erichkeane replied:
> On the other hand SYCL is a trademark from Khronos, so if it can stay SYCL it is better from this perspective. Otherwise people get confused and create more effort like #113060 😃

I retract that point anyway, I cannot find where I got that from, it doesn't seem to be in our coding standard, though I thought it was!  So I guess that was just an N-AI (not-artificial intelligence... or Artificial Not-intelligence?  you choose...) hallucination.

tahonermann replied:
With regard to the `Is`, I was just preserving the style that was already there. The corresponding command line options are named `-fsycl-is-device` and `-fsycl-is-host`, so perhaps it is worth maintaining that symmetry. I'm ambivalent.

erichkeane replied:
Those are a weird choice too IMO.  I think we inherited that without much review from the downstream, but I kind of want to push back.  The `-is` doesn't really buy us anything here.  I'd accept the flag as an alias, though I'm not sure how much value there is as we aren't really functional in the upstream, so our users are limited.

tahonermann replied:
I've never loved those names either. I don't think it makes sense to change them as part of this PR though. https://github.com/llvm/llvm-project/pull/107493 is introducing new `-fsycl-device-only` and `-fsycl-host-only` options as aliases of `--offload-device-only` and `--offload-host-only` (also aliased by `--cuda-device-only` and `--cuda-host-only`). It might make sense to argue for a change in that PR. Perhaps there is an opportunity to remove the `-fsycl-is-device` and `-fsycl-is-host` options in favor of a common offload indicator of host vs device compilation.

For the purposes of this PR, I think it makes sense to stick with the status quo pending a proposal for a better, perhaps unified, way of handling host vs device compilation.

erichkeane replied:
> I've never loved those names either. I don't think it makes sense to change them as part of this PR though. #107493 is introducing new `-fsycl-device-only` and `-fsycl-host-only` options as aliases of `--offload-device-only` and `--offload-host-only` (also aliased by `--cuda-device-only` and `--cuda-host-only`). It might make sense to argue for a change in that PR. Perhaps there is an opportunity to remove the `-fsycl-is-device` and `-fsycl-is-host` options in favor of a common offload indicator of host vs device compilation.
> 
> For the purposes of this PR, I think it makes sense to stick with the status quo pending a proposal for a better, perhaps unified, way of handling host vs device compilation.

Please lobby this there.  I'm OK with leaving it in the flag name here, but I still would like the lang-opt changed here.

tahonermann replied:
The `SYCLIsDevice` and `SYCLIsHost` members of `LangOpt` are pre-existing and referenced from numerous source files. I'm not going to change them as part of this PR. You are welcome to submit a PR to change them if you like. Otherwise, we can investigate changing them as part of a future offload unification effort.

erichkeane replied:
To me, this is a "longer we wait, the harder this is". So no time like the present!

AaronBallman replied:
IMO, where we eventually want to end up is with `IsDevice` and `IsHost` which composes with all the offloading languages (I cannot imagine a situation where you're compiling for the device in one offloading language while compiling for the host in another offloading language in the same compilation pass). e.g., `getLangOpts().SYCL && getLangOpts().IsDevice` tells you that you're doing a device compilation for SYCL.

erichkeane replied:
Aaron chatted with me off-line, and I think THAT is the eventual place we want to be.  So doing this work now, while beneficial, would eventually become wasted work.  So I'm resolving this.

tahonermann replied:
Thanks, Erich. The model Aaron described matches what I had in mind as part of some future offload unification effort, so it sounds like we are all aligned here.

tahonermann commented on clang/include/clang/AST/SYCLKernelInfo.h:
```diff
@@ -0,0 +1,42 @@
+
+#include "clang/AST/Decl.h"
+#include "clang/AST/Type.h"
+#include <string>
```
Nothing right now. It will be needed for later changes; I neglected to remove it when preparing this patch. I'll remove it.

tahonermann replied:
Done, thanks for catching this.

tahonermann commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,174 @@ The SYCL kernel in the previous code sample meets these expectations.
+following requirements.
+
+* Has a ``void`` return type.
+* Is not a non-static member function, constructor, or destructor.
```
That would be a return to previous text that you didn't care for, "is a static member function or non-member function". Most of these requirements are a list of prohibitions (the `void` return type requirement is an exception). I tend to prefer a concise list of what is not allowed; it tends to map better to diagnostics. I have a slight preference to leave this as is, but not strong opinion.

erichkeane replied:
Ah, it was long enough I forgot that.  Ok, I'm fine leaving this how it is.  Curious if we find ourselves having to update this list in teh future.

tahonermann commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,174 @@ The SYCL kernel in the previous code sample meets these expectations.
+* Is not a C variadic function.
+* Is not a coroutine.
+* Is not defined as deleted or as defaulted.
+* Is not declared with the ``constexpr`` or ``consteval`` specifiers.
```
`constexpr` would only be meaningful here if the SYCL invocation functions like `single_task` were also declared `constexpr`. I think this would be worth reconsidering when/if an effort appears to enable `constexpr` support in the SYCL library.

erichkeane replied:
I don't see a good reason to limit it here, I think we should just implement it as support.  Particularly as the library isn't necessarily the only user of this.  Plus, it allows users to USE existing library functions.  That is, it is an artificial restriction on users who wnat to re-use existing code, even if we don't ever execute at constexpr time.

tahonermann replied:
I think there is some misunderstanding here. Can you provide an example that illustrates what you have in mind?

These functions aren't useful on their own; a call to them is specified as a no-op and is done for the side effect of triggering emission of a kernel entry point function (any ODR-use will do for that; it doesn't have to be a call). I think it would be quite strange if a call in a constexpr context evaluated the body of the function but a call outside did not. From the documentation in the PR:
> The call to ``kernel_entry_point()`` has no effect other than to trigger emission of the entry point function. The statments that make up the body of the function are not executed when the function is called; they are only used in the generation of the entry point function.

erichkeane replied:
The body of the `kernel_entry_point` is still required to be valid C++, and often just call the function, unless they are replaced at code-gen time (at which point, what you quote is correct).  

We'd received quite a few bugs/reports/wishes that these functions could be called in a constexpr context such that their body would just be executed in that context (since the body actually DOES the correct evaluation).  One such use case was folks trying to not use the SYCL library, but instead just augment their existing program to get 'free' parallelism, (plus some alterations because of our limitations on the shape of the function) and then being disappointed because they actually have valid constexpr uses.

If the SYCL library doesn't choose to do this (and the onus is on the author of the function to ensure that the 'as-if' rule gets satisfied by the body) they don't have to, but the limit here is artificial IMO.

keryell replied:
Just rethinking about this, since I have worked today on `constexpr` in a related context https://github.com/KhronosGroup/SYCL-CTS/pull/976
This looks like a pessimistic interpretation of the specification or something we could clarify in the SYCL committee.
It is not really important to consider `consteval` as it will always be in a constant evaluated context, so it will never reach SYCL codegen (perhaps in the future if we want to speed up the slooow constant evaluator of Clang itself on some accelerators? :rofl: ).
And why not having a kernel which is `constexpr` function too? It will only matter when in a non-constant evaluated context.

erichkeane replied:
>And why not having a kernel which is constexpr function too? It will only matter when in a non-constant evaluated context.

That is sorta my thoughts too.  `consteval` doesn't really make sense:  We COULD do it, but it would be a no-op attribute, which I think is reason enough to diagnose it.  

`constexpr` on the other hand definitely makes sense to me.

tahonermann replied:
With regard to `constexpr`, consider the following example. Is the lambda actually invoked when `f()` is called?
```
template<typename KN, typename KT>
[[clang::sycl_kernel_entry_point(KN)]]
void skep(KT k) { k(); }
void f() {
  skep<struct K>([]{});
}
```

The answer is no. As described in the documentation and quoted in an earlier comment in this conversation, the call to the `sycl_kernel_entry_point` attributed function happens, but it's apparent body is not actually executed. The body serves as a pattern for the offload kernel entry point function.

With that in mind, what should the following do?
```
constexpr auto v = (skep<struct K>([]{}), 0);
```

The answer is that it should do nothing (invoking the lambda would be inconsistent with what happens outside of constant evaluation).

The way to support constant evaluation of a SYCL kernel (which may be `constexpr`) would be to write the SYCL invocation function to detect constant evaluation and react accordingly.
```
template<typename KN, typename KT>
constexpr void single_task(KT k) {
  if consteval {
    k();
  } else {
    skep<KN>(k); // Trigger emission of the offload kernel entry point.
    // FIXME: Do stuff to invoke the offload kernel entry point
  }
}
constexpr auto v = (single_task<struct K>([]{}), 0); // Huzzah, constant evaluated SYCL kernel.
```

erichkeane replied:
That is what I was talking about that the author of said function ends up being responsible for the as-if rule.

tahonermann replied:
Thanks, Erich. I'd still like to understand what your mental model is for what `constexpr` would mean on one of these functions; just to ensure there isn't some misunderstanding.

I've been assuming that the model you had in mind would effectively implicitly inline the `if consteval` statement from my `single_task()` example into `skep()` such that it would behave when called as-if written as follows:
```
template<typename KN, typename KT>
[[clang::sycl_kernel_entry_point(KN)]]
void skep(KT k) {
  if consteval {
    k();
  } else {
    // Do nothing; an offload entry point function will have been generated
    // that the SYCL run-time library will arrange to call.
  }
}
```

This is the reason for my comments about inconsistent behavior in and out of constant evaluation context. Note that any caller would still have to be written to be sensitive to whether invocation is occurring during constant evaluation since it would have to arrange (or not) for the offload entry point to be invoked on the appropriate device.

If you have a different model in mind, please do share details.

erichkeane replied:
That is effectively  my mental model, except without the `if consteval`, since the body of the `skep` doesn't really exist when doing the offload codegen. The point is there is a lot of `device` stuff that gets arranged that is, in my mental model, implementation details for "call this function object, just faster/more parallel". 

However, the 'faster/more parallel' is a bit of a QoI thing, so 'call this slower, but at compile time' should work as well.

tahonermann replied:
Ok, thanks. Note that this design doesn't require the `#ifdef __SYCL_DEVICE_ONLY__` dance that was needed with the `sycl_kernel` attribute. This is an intentional divergence made to accommodate a future single pass compilation model. So, the body of `skep()` does exist during both host and device compilation, it just isn't evaluated when called. These are details for a later PR.

> However, the 'faster/more parallel' is a bit of a QoI thing, so 'call this slower, but at compile time' should work as well.

I agree, but that needs to be handled at a higher level. `sycl_kernel_entry_point` attributed functions can't fill the role of a user facing interface since they don't include the mechanism needed for device dispatch. These functions can only be used as part of the implementation detail for higher level interfaces.

keryell replied:
After some background thoughts, this `constexpr` aspect is only for the runtime function itself and more generally the SYCL runtime is currently far from being `constexpr`. So, we can just prevent `constexpr` and `consteval` and change this implementation detail in the future if this becomes a problem.
Note that this does not prevent to have the kernel lambda to be `constexpr` today.

tahonermann replied:
Thank you, @keryell, yes, that matches my perspective. I'll go ahead and resolve this conversation now. If you have no further concerns, could you please approve this PR?

tahonermann commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,174 @@ The SYCL kernel in the previous code sample meets these expectations.
+* Is not a coroutine.
+* Is not defined as deleted or as defaulted.
+* Is not declared with the ``constexpr`` or ``consteval`` specifiers.
+* Is not declared with the ``[[noreturn]]`` attribute.
```
No, we dropped the function template requirement as being unnecessary. This permits an implementation to declare single purpose kernels if desired.

erichkeane replied:
Great!

tahonermann commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -14296,6 +14296,29 @@ void ASTContext::getFunctionFeatureMap(llvm::StringMap<bool> &FeatureMap,
   }
 }
 
+static SYCLKernelInfo BuildSYCLKernelInfo(CanQualType KernelNameType,
```
I agree it is silly as is. It will get bigger; see [here](https://github.com/tahonermann/llvm-project/compare/sycl-upstream-fe-pr3...tahonermann:llvm-project:sycl-upstream-fe-pr4#diff-328fcdcf882c121d5c56fdc6cd56c43fdbbb00da171da70f3ba6982c43686a63R14315-R14357).

erichkeane replied:
Got it, thanks!

tahonermann commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -12053,6 +12053,10 @@ bool Sema::CheckFunctionDeclaration(Scope *S, FunctionDecl *NewFD,
     OpenMP().ActOnFinishedFunctionDefinitionInOpenMPAssumeScope(NewFD);
 
+  if (LangOpts.isSYCL() && NewFD->hasAttr<SYCLKernelEntryPointAttr>() &&
+      !NewFD->isInvalidDecl() && !NewFD->isDependentContext())
```
Hmm, I'm not a fan of interfaces that implicitly do nothing under some set of conditions; I prefer that callers be required to consider whether those conditions require some other action.

erichkeane replied:
I've been thinking the opposite lately, which is why I made this comment.  These conditions end up getting 'forgotten' somewhere along the way, and we get something that acts weird in release build.

tahonermann replied:
That is exactly my concern and why I want such mistakes to be caught at compile-time. This is the motivation for ensuring that an incorrect registration fails compilation even if a check or diagnostic is missed somewhere.

erichkeane replied:
At this point though, you're not actually catching anything at compile-time.  All you're doing is asserting at runtime if someone forgets one of these tags, rather than just preventing us from having to worry about it.

tahonermann replied:
The call to `llvm::report_fatal_error()` will occur at run-time for Clang, sure, but that will be compile-time for users building their SYCL source code. I would rather have Clang fail with a stack trace that clearly indicates the problem (to us) then perform code gen thinking that a kernel was registered when one wasn't (likely to manifest in some kind of "kernel not found" error at run-time) or, worse, having a duplicate registration silently rejected such that an unintended kernel gets called. I grant that these scenarios will only occur if there are missing or incorrect checks elsewhere, but these kinds of problems would be rather painful to trace back to what the problem really is.

erichkeane replied:
And I'm saying that is SUCH a divergence from the standard behavior of the compiler and our current coding standards, that I don't want to do it the first time here.

We likely want some sort of RFC for that which audits all current asserts/unreachables/etc, and figures out which of those should be `report_fatal_error`, and a comprehensive patch to fix all/most of them at once.

tahonermann replied:
This is not the first such case. A few existing examples are listed below. More examples can be found in Serialization.
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/AST/ExternalASTSource.cpp#L121
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/Basic/ProfileList.cpp#L62
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/Basic/SanitizerSpecialCaseList.cpp#L36
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/Basic/SourceManager.cpp#L674
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/CodeGen/CoverageMappingGen.cpp#L1030-L1031
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/CodeGen/CoverageMappingGen.cpp#L1050-L1051
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/CodeGen/LinkInModulesPass.cpp#L30
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/CodeGen/ObjectFilePCHContainerWriter.cpp#L277
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/Interpreter/Interpreter.cpp#L426-L428

erichkeane replied:
I don't doubt that others got through review, and a few of those I clicked on are different enough situations as to not be the same as this. This is a novel use of that function, and one caught during review.

tahonermann replied:
I think this is substantially similar. It seems we aren't going to agree. Let's see what others have to say.

tahonermann replied:
I'm going to resolve this conversation about the use of `llvm::report_fatal_error()` since the most recent comments on this topic are happening in https://github.com/llvm/llvm-project/pull/111389#discussion_r1825122665.

tahonermann commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,174 @@ The SYCL kernel in the previous code sample meets these expectations.
+#. Reconstituting the SYCL kernel object, if necessary, using the offload
+   kernel parameters.
+
+#. Calling the ``operator()`` member function of the (reconstituted) SYCL kernel
```
The call to the `operator()` function is resolved when the `sycl_kernel_entry_point` attributed function is compiled. The SYCL kernel entry point function just calls that same function; it is generated from the same AST (lightly transformed).

erichkeane replied:
Got it, thanks.

tahonermann commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -455,6 +455,174 @@ The SYCL kernel in the previous code sample meets these expectations.
+of these data members corresponds to a special SYCL type that must be passed
+individually as an offload kernel parameter, it is necessary to decompose the
+SYCL kernel object into its constituent parts; the offload kernel will have
+two kernel parameters. Given a SYCL implementation that uses a
```
Yes, that is a possibility that depends on the particulars of how `sycl::stream` is defined. I refrained from going into further detail about such decomposition for now since we don't yet have a PR for how SYCL special types will be handled. Support for SYCL special types will likely involve additional new attributes and we can document their special behavior then.

erichkeane replied:
Ok, then perhaps some slight clarity, you mention 'decomposing', but then imply that decomposing only results in 1 argument here, when it could be 1 or more.  Perhaps something about `the offload kernel will have two or more kernel parameters (as the decomposition of the ``stream`` can result in multiple parameters)`.

WDYT?

tahonermann replied:
I went with an approach of adding an additional entry to the notes that follow the example.
> The depiction of the ``sycl::stream`` parameter as a single self contained kernel parameter is an oversimplification. SYCL special types may require additional decomposition such that the generated function might have three or more parameters depending on how the SYCL library implementation defines these types.

erichkeane reviewed (changes requested):
See other comments/discussion.

tahonermann committed:
Round two of addressing code review feedback from Erich Keane.

Changes include:
- Removal of an unneeded inclusion of <string>.
- An update to the `sycl_kernel_entry_point` attribute documentation.
- Updates to the AST test to validate next line proximity where applicable.

tahonermann committed:
Round three of addressing code review feedback from Erich Keane.

Changes include:
- Use `isTemplated()` instead of `isDependentContext()` since the latter doesn't
  necessarily handle templated friend functions.

tahonermann committed:
Round four of addressing code review feedback from Erich Keane.

Changes include:
- Replaced a use of `llvm::report_fatal_error()` with `assert()`.
- Modified `registerSYCLEntryPointFunction()` to silently skip registration
  of invalid or dependent function declarations.

tahonermann commented:
@erichkeane, I believe all of the concerns you raised have been resolved with the exception of the `constexpr` question for which you asked @keryell to weigh in. I hope we're close to consensus on no need for a change for that one. Please re-review when time permits.

erichkeane commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -14296,6 +14296,34 @@ void ASTContext::getFunctionFeatureMap(llvm::StringMap<bool> &FeatureMap,
+  CanQualType KernelNameType = getCanonicalType(SKEPAttr->getKernelName());
+  auto IT = SYCLKernels.find(KernelNameType);
+  if (IT != SYCLKernels.end()) {
+    assert(declaresSameEntity(FD, IT->second.getKernelEntryPointDecl()) &&
```
```suggestion
    assert((IT == SYCLKernels.end() || declaresSameEntity(FD, IT->second.getKernelEntryPointDecl())) &&
```

Then just kill the whole branch.

erichkeane reviewed (approved):
1 nit, else LGTM.

Please let @keryell chime in on the constexpr decision.

tahonermann committed:
Round five of addressing code review feedback from Erich Keane.

Changes include:
- Minor refactoring to eliminate a branch that only contained an assertion.

keryell reviewed (approved):
Thank you!
